j5/backends/hardware/sr/v4/serial/protocol.py

```diff
@@ -0,0 +1,184 @@
+
+from j5.backends import CommunicationError
+from j5.backends.hardware.j5.serial import SerialHardwareBackend
+
+
+class BoardIdentity(NamedTuple):
+    """Identifiers for a Student Robotics v4 board."""
+
+    vendor: str
+    board: str
+    asset_tag: str
+    software_version: str
+
+
+class BoardVersion(NamedTuple):
+    """The version returned from a board."""
+
+    hw_ver: int
+    fw_major: int
+    fw_minor: int = 0
+
+
+class SRV4SerialProtocolBackend(SerialHardwareBackend, ABC):
+    """
+    Backend for a Student Robotics v4 board using the serial protocol.
+
+    The protocol is loosely based on SCPI.
+    """
+
+    _lock: threading.Lock
+
+    @property
+    def firmware_version(self) -> str:
+        """
+        The firmware version reported by the board.
+
+        :returns: firmware version reported by the board, if any.
+        """
+        identity = self.get_identity()
+        return identity.software_version
+
+    def get_identity(self) -> BoardIdentity:
+        """
+        Get the board identity information.
+
+        :raises CommunicationError: Response did not match the expected format.
+        :returns: A tuple of board identity information
+        """
+        response = self.query("*IDN?")
+        parts = response.split(":", 4)
+        if len(parts) != 4:
+            raise CommunicationError(
+                f"Identify response did not match format: {response}")
+        else:
+            return BoardIdentity(*parts)
+
+    def reset_board(self) -> None:
+        """
+        Resets the board.
+
+        This turns off all the board's outputs, LEDs, etc.
+        """
+        self.request("*RESET")
+
+    def request(self, command: str) -> None:
+        """
+        Sends a request to the board.
+
+        A request command does not return any data and will respond
+        ACK if the request was executed successfully.
+
+        :param command: The command to be sent as part of the request.
+        :raises CommunicationError: Failed to send request.
+        """
+        response = self._raw_request(command)
+
+        if response != "ACK":
+            raise CommunicationError(
+                "Expected ACK from Power Board, but got: {response}",
+            )
+
+    def query(self, command: str) -> str:
+        """
+        Sends a query to the board and returns its response as a string.
+
+        This method expects a response with additional data
+        and will raise an exception if only an ACK is received.
+
+        Query commands end with '?'.
+
+        :param command: The command to be sent as part of the query.
+        :returns: Response.
+        :raises CommunicationError: Failed to send request or only ACK is returned.
+        """
+        response = self._raw_request(command)
+        if response and response != "ACK":
+            return response
+        else:
+            raise CommunicationError("Power board responded with ACK but expected data.")
```

Comment:
ACK is still a response, we can leave handling responses to the caller.

Reply:
Given that the caller in this instance is still within the class, I would rather handle this case once rather than many times.